### ISSUE TITLE
5x throughput drop during terminate and replace

### ISSUE THREAD

**Issue by bentsi** [scylladb/scylla#4712]
*Installation details*
Scylla version (or git commit hash):  3.1.0.rc2-0.20190713.92bf92817
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI):  ami-03d6caf668ad59911

The issue happens when streaming and sending hints start to the new node 

![image](https://user-images.githubusercontent.com/5269241/61220516-22bac100-a71f-11e9-81dd-1da530cdb6c5.png)

Grafana snapshot:
https://snapshot.raintank.io/dashboard/snapshot/xiZ7RageeHJFiCgBYtkddsA2b5VzZXS1

**Comment by piodul** [comment]
This issue is being caused by hints overloading the cluster. In this situation, in the time between old node being shut down and new one becoming UN, a number of hints have been accumulated. Hints are segregated separately, with respect to their target endpoint. Because the target node has left the cluster, the hints manager for that endpoint is drained, which means hints will be sent before removing the hints directory.

Obviously, those hints cannot be sent to their intended recipient. Therefore, current logic sends them with `cl=ALL` to all its current replicas of each hint, as evidenced in the function below:

```cpp
future<> manager::end_point_hints_manager::sender::do_send_one_mutation(frozen_mutation_and_schema m, const std::vector<gms::inet_address>& natural_endpoints) noexcept {
    return futurize_apply([this, m = std::move(m), &natural_endpoints] () mutable -> future<> {
        // The fact that we send with CL::ALL in both cases below ensures that new hints are not going
        // to be generated as a result of hints sending.
        if (boost::range::find(natural_endpoints, end_point_key()) != natural_endpoints.end()) {
            manager_logger.trace("Sending directly to {}", end_point_key());
            return _proxy.send_hint_to_endpoint(std::move(m), end_point_key());
        } else {
            manager_logger.trace("Endpoints set has changed and {} is no longer a replica. Mutating from scratch...", end_point_key());
            // FIXME: using 1h as infinite timeout. If a node is down, we should get an
            // unavailable exception.
            auto timeout = db::timeout_clock::now() + 1h;
            //FIXME: Add required frozen_mutation overloads
            return _proxy.mutate({m.fm.unfreeze(m.s)}, consistency_level::ALL, timeout, nullptr, empty_service_permit());
        }
    });
}
```

One solution I have come up with is to send hints on a separate channel in this case as well. It still causes a drop in performance, but not as big as with current logic.

I reproduced a similar scenario to the one that happened in the test: stop one node, replace it with a new one. Additionally, I ran `nodetool repair` after the new node became UN. Here are relevant graphs for the implementation using separate connection:

![2020-01-20-121602_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72722677-338d3900-3b7f-11ea-8f66-3426cf86a91a.png)

![2020-01-20-121620_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72722685-3ab44700-3b7f-11ea-9f67-12ddd3658d36.png)

Another option would be to calculate the new target in a smart way. To see how it would fare performance-wise, I prepared a dumb implementation that sends orphaned hints to a hardcoded node, and tested it. In this case I didn't run `nodetool repair`, though. The performance is even better:

![2020-01-20-121748_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72722694-40aa2800-3b7f-11ea-9768-438fca0d6b9f.png)

![2020-01-20-121829_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72722706-47389f80-3b7f-11ea-9ca2-61e8f8f86a8c.png)

In all graphs above a following observation can be made: after nearly all accumulated hints are sent, the request throughput goes up, but arrives and stays at a level that is worse than before the old node was terminated. I have a theory why that happens, and I wrote about it [here](https://github.com/scylladb/scylla/issues/4981#issuecomment-576226763).

**Comment by slivne** [comment]
@piodul - I am missing something here - I don't understand the original code CL:ALL assumption here - does C* do the same ?

Do they have an open bug about it. 

While sending to CL:ALL compared to CL:ONE will require more work - maybe the issue is the pace in which we are sending those requests - and that needs to be throttled down - are we limiting this in any way.

**Comment by piodul** [comment]
@slivne 

When hint's destination endpoint is no longer its replica, we don't make any assumptions which endpoint should receive it instead, we just sent it to all its current replicas. When issuing a write, if CL is achieved but some of the replicas is unavailable, we store a hint for that replica. If weaker CL than ALL would be used, then we would risk generating more hints for unavailable replicas.

I looked into Cassandra's source (master branch), and noticed that its behavior is slightly different. When a hint is replayed and its destination is no longer its replica, the hint is transformed into multiple hints with destinations being set to current replicas.

There are some more differences, related to this issue:

- When a node is being replaced (the scenario from this issue), Cassandra just... forgets about these hints. Hints for the node from before the replacement are neither sent anywhere, nor deleted. A periodic job of the hint manager is still being run, but it never decides to do anything with them. It looks like an omission on their part, but I couldn't find any issues on their bug tracker related to hints being left and forgotten.
- When a node is being decommissioned (nodetool decommission), it transfers all hints it had stored to the closest live node. The destination node stores them and assumes responsibility of sending them to their destination.
- When a node is being removed (nodetool removenode), other nodes delete hints targeted at that node.

As for limiting send rate, we already limit the amount of hints being sent simultaneously - 128 hints per shard, or 10% of the total memory. Cassandra limits the amount of memory occupied by simultaneously sent hints by a configuration parameter. This parameter is scaled by the number of other nodes in the cluster.

The PR #5668 aims to reduce the pace in which hints are handled. Using a separate connection in this case allows us to handle writing hints on a different scheduling group, and avoid problems related to hints sharing the same tcp queue as regular mutations. We already do this for non-orphaned hints. In my manual tests on AWS, this solution worked quite well (I'll run the same SCT scenario as in the issue soon to properly compare the difference).

Nevertheless, we have an option to try to be smarter in this scenario, and try to send the orphaned hints only to the new node.

@vladzcloudius @gleb-cloudius what do you think, what solution would be best?

**Comment by avikivity** [comment]
If we have a hint for a node which is being range-moved, perhaps we should send to both the old and the new owners, in order to ensure that we end up with the hint applied regardless of whether the range movement succeeded or not.

**Comment by avikivity** [comment]
Of course after the range movement succeeds, we lose this information and we'd end up sending to the wrong node.

**Comment by vladzcloudius** [comment]
@piodul I think the current solution (send to all current replicas) is the best because it's the simplest. This way you don't need to calculate what's the new replica and what's not (like in @avikivity example above). The same code works regardless on what's the cluster's current state topology wise.

The impact of all these hints being sent  IS an issue but I believe that this issue is going to be handled by PR #5668. 

Can we validate it?

**Comment by piodul** [comment]
I'll try re-running the test with the patch from #5668. @bentsi, this was a longevity test, right? Do you remember which configuration was used here?

**Comment by bentsi** [comment]
@piodul Test YAML was something similar to: `test-cases/longevity/longevity-2TB-4days-authorization-and-tls-ssl-1dis-2nondis-nemesis.yaml`, but this scenario is very complex, it runs multiple random nemesis in parallel. I suggest to use this scenario with one relevant Nemesis only (terminate and replace).
To achieve this you will need to create your conf YAML based on the above and change `nemesis_class_name: 'NodeTerminateAndReplace'`. Let me know if you need help to do this

**Comment by piodul** [comment]
I started a longevity job with configuration adjusted to only use the `NodeTerminateAndReplace` nemesis: [link](https://jenkins.scylladb.com/view/master/job/scylla-master/job/longevity/job/byo-longevity-test/42/). I'll update when I observe a nemesis run.

**Comment by piodul** [comment]
The longevity job performed one node replacement. Here are the artifacts:

```
+--------------------------------------------------------------------------------------------------------------------------------------------+
|           Collected logs by test-id: 5e1b3781-8084-423e-8261-d8f44e0153a2 Directory: /sct/20200206-070055-735319/collected_logs            |
+-------------+------------------------------------------------------------------------------------------------------------------------------+
| Cluster set | Link                                                                                                                         |
+-------------+------------------------------------------------------------------------------------------------------------------------------+
| db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/5e1b3781-8084-423e-8261-d8f44e0153a2/20200207_155317/db-cluster-5e1b3781.zip  |
| monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/5e1b3781-8084-423e-8261-d8f44e0153a2/20200207_155317/monitor-set-5e1b3781.zip |
| loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/5e1b3781-8084-423e-8261-d8f44e0153a2/20200207_155317/loader-set-5e1b3781.zip  |
| sct-runner  | https://cloudius-jenkins-test.s3.amazonaws.com/5e1b3781-8084-423e-8261-d8f44e0153a2/20200207_155317/sct-runner-5e1b3781.zip  |
+-------------+------------------------------------------------------------------------------------------------------------------------------+

```

There is also a live preview: [here](http://54.211.1.4:3000/d/5RXV2o8Wz/scylla-per-server-metrics-nemesis-master?orgId=1&from=1581045660744&to=1581074584252&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All)
Or you can download and browse locally with `hydra investigate show-monitor 5e1b3781-8084-423e-8261-d8f44e0153a2`

The longevity job performed one replacement operation successfully. Comparing graphs from the issue description and from this run, the cluster handled node replacement better:

![total-requests](https://user-images.githubusercontent.com/4950936/74138156-5e543580-4bf1-11ea-848d-9483e47b3d33.png)

While there still is a throughput drop when the new node becomes UN, it is much shorter (around 12 minutes in this case) and has a different shape - originally, throughput was recovering gradually over time, here it jumps up to around 80% of the previous performance, which is expected considering hints and active repair operation.

Trying to determine the cause of this drop, I looked at the amount of tasks in each scheduling group, and noticed a huge increase in the compaction group, around the time of when the drop happened:

![scheduling-groups](https://user-images.githubusercontent.com/4950936/74138194-6f9d4200-4bf1-11ea-872f-8ed943082601.png)

----

There was another attempt to replace a node, but one of the nodes dumped core, cassandra-stress processes exited due to insufficient consistency level, and I had to abort the test. The core dump can be found [here](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.996.e4505954431d4211adb57405513d9152.7934.1581079032000000/core.scylla.996.e4505954431d4211adb57405513d9152.7934.1581079032000000.gz)

I did not dig in the core dump. The decoded stack trace is as follows:

```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:56
seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:722
 (inlined by) print_with_backtrace at ./build/release/seastar/./seastar/src/core/reactor.cc:743
seastar::print_with_backtrace(char const*) at ./build/release/seastar/./seastar/src/core/reactor.cc:750
seastar::install_oneshot_signal_handler<6, &seastar::sigabrt_action>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) at ./build/release/seastar/./seastar/src/core/reactor.cc:3436
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3418
 (inlined by) _FUN at ./build/release/seastar/./seastar/src/core/reactor.cc:3414
?? ??:0
addr2line: '/opt/scylladb/libreloc/libc.so.6': No such file
/opt/scylladb/libreloc/libc.so.6 0x000000000003c624
/opt/scylladb/libreloc/libc.so.6 0x00000000000258d8
seastar::throw_system_error_on(bool, char const*) at ./build/release/seastar/./seastar/include/seastar/core/posix.hh:423
 (inlined by) seastar::file_desc::sendmsg(msghdr const*, int) at ./build/release/seastar/./seastar/include/seastar/core/posix.hh:257
 (inlined by) seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}::operator()() at ./build/release/seastar/./seastar/src/core/reactor.cc:370
 (inlined by) seastar::apply_helper<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply({lambda()#1}&&, std::tuple<>) at ./build/release/seastar/./seastar/include/seastar/core/apply.hh:36
 (inlined by) auto seastar::apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}>(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&, std::tuple<>&&) at ./build/release/seastar/./seastar/include/seastar/core/apply.hh:44
 (inlined by) seastar::future<unsigned long> seastar::futurize<seastar::future<unsigned long> >::apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}>(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&, std::tuple<>&&) at ./build/release/seastar/./seastar/include/seastar/core/future.hh:1645
 (inlined by) seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<>&&)#1}::operator()(seastar::future_state) at ./build/release/seastar/./seastar/include/seastar/core/future.hh:1191
 (inlined by) seastar::continuation<seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<>&&)#1}>::run_and_dispose() at ./build/release/seastar/./seastar/include/seastar/core/future.hh:506
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2130
seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2553
seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2536
 (inlined by) seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2698
seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const at ./build/release/seastar/./seastar/src/core/reactor.cc:3838
std::function<void ()>::operator()() const at /usr/include/c++/9/bits/std_function.h:690
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:52
addr2line: '/opt/scylladb/libreloc/libpthread.so.0': No such file
/opt/scylladb/libreloc/libpthread.so.0 0x00000000000094e1
/opt/scylladb/libreloc/libc.so.6 0x00000000001016d2
```

**Comment by vladzcloudius** [comment]
Interesting results. 
Here is another interesting angle (note that I zoomed into the interesting time segment where there was that dive):

Notice that the CQL dive continues even after the initial hints spike is over.

![image](https://user-images.githubusercontent.com/6103981/74179827-30c8b580-4c0c-11ea-9711-7729cfeebc41.png)

**Comment by slivne** [comment]
@piodul what is the expected status now after you patch to move the hints to the correct verb ?

**Comment by piodul** [comment]
I dug into logs and metrics from the last SCT run, and have more information now. For context, here are graphs with metrics I felt were relevant.

![zoomed-in](https://user-images.githubusercontent.com/4950936/77546934-d4181600-6eac-11ea-86a0-61bb4461787b.png)

Zoomed out:

![zoomed-out](https://user-images.githubusercontent.com/4950936/77546957-d9756080-6eac-11ea-9303-de33951b8305.png)

There were repairs, which operated on the following keyspaces, in that order: `system_traces`, `mview`, `standard1`, `system_auth`. On the graphs above, I marked moments when one repair stopped and/or the next repair started. The last repair was very short, so the fourth line (visible on second image only) represents start and end of the last repair.

I'll focus mainly on repairs for `mview` and `standard1`, as those repairs took the most time. Here are some key points:

- Keyspace `mview` contained one table and two materialized views based on that table, while `standard1` did not contain any MVs.
- During repair of `mview`, performance dropped to around 14K requests/s, while during repair of `standard1` it was around 35K requests/s. Before repairs, it was at the level of 42K requests/s.
- The number of compactions increased greatly during the repair of `mview` - the new node had around 50-60 compactions, other nodes had around 200-400 compactions. Before repairs, each node had about 12 compactions active.
- During repairs, because of the node replacement, all MV and non-MV hints accumulated originally for the node that was killed were sent with RF=3 to all their current replicas.
- Hint sending was active both during repair of `mview` and `standard1`, but hint sending rate seemed to be slightly bigger and more consistent during repair of `mview`.

I also looked at the view update backlog, thinking that it might have caused backpressure of the regular writes, but I do not see a correlation between its size and request throughput or any other metric.

@psarna, @asias Could you take a look? Do you have any idea why the increased number of compactions could occur when repair, materialized views and hints happen at the same time?

**Comment by piodul** [comment]
> @piodul what is the expected status now after you patch to move the hints to the correct verb ?

@slivne I ran the test suggested by Bentsi with Scylla version that had the fix, and a performance drop still occured. Now, the criteria seem to be more specific - the drop happened only during repair of a keyspace that contained materialized views, otherwise the performance was reasonable. I pinged Piotr S and Asias, and will investigate/discuss further.

**Comment by psarna** [comment]
@piodul do you also have a way of determining what was the number of sstables?

Materialized views causing trouble during repair sound like an issue with staging sstables - i.e. sstables received from another node, which are temporarily not subject to compaction, because we wait until we generate view updates from them. We don't generate view updates from all incoming sstables, but in case of repair we do - so every received sstable will first wait in the staging queue, and during that period it will be excluded from compactions. Now, the question is why having staging sstables also results in a storm of compactions, while it should be the other way round - these sstables are explicitly not registered for compaction:
https://github.com/scylladb/scylla/blob/eca0ac5772c72c2f53671dce63284d36e25473c8/table.cc#L697-L701

They are later registered for compaction, after view updates are properly generated from them, via `table::move_sstables_from_staging`, but that also shouldn't cause a 5x raise in the *number* of compactions... does this number vary between compaction strategies? Maybe that would give more context, otherwise I don't have a clear idea what could be the cause.

**Comment by slivne** [comment]
@piodul / @psarna  the reason you are seeing repair now is that we moved to repair based operations (that is we will do repairs in some operations instead of streaming)

There is an issue related to repair based operations and materialized views that Asias has a fix
https://github.com/scylladb/scylla/pull/6000 

So don;t focus on the matrialized views / secondary indexes - we know why this is "worse" then whats expected and it will be fixed.

If its simple to split the regular tables apart do that if not - wait for the fix togo in and then retest

**Comment by vladzcloudius** [comment]
@slivne @psarna The MV backlog @piodul is seeing with a very high probablitlity may be related to https://github.com/scylladb/scylla/issues/6069

**Comment by piodul** [comment]
I'm running the test again, but this time I removed cassandra-stress process that writes to MV tables. I'll update after the test completes.

[Jenkins link](https://jenkins.scylladb.com/view/master/job/scylla-master/job/longevity/job/byo-longevity-test/57/)

**Comment by piodul** [comment]
The byo job ran the nemesis a few times already, so I aborted it. Live metrics are here: [link](http://18.206.159.72:3000/)

The test scenario was the same as before, with the exception that this time there was no cassandra-stress process that caused materialized view updates. The TerminateAndReplace nemesis had 6 complete runs, but I'll concentrate on the first - there were unrelated issues that could interfere with the analysis (one node dumped core, metrics didn't seem to be collected from some new nodes).

Results look much better now:

![terminate-and-replace-metrics](https://user-images.githubusercontent.com/4950936/77911479-25e7e400-7291-11ea-8098-6c4554429baf.png)

A small note: as I stated in the previous analysis, the lowest point in the "Total requests" graph is a lie - at the moment merked by blue vertical lines, prometheus was reconfigured to collect metrics from the new node, so the height of "Total requests" graph to the left of the line should be 25% higher - more like the height immediately to the right.

Now, when hints are being sent, throughput is around 24K reqs/s. After all hints are sent, throughput is at around 32K reqs/s. So, the performance drops to about 75% during hint sending, which is a much better result than in the original test.

**Comment by piodul** [comment]
@slivne @vladzcloudius Could you take a look at the new results above? I removed tables with MVs from the test and now performance decreases only by about 20-25% when hints are involved. Is this acceptable? If so, can we close this issue?

**Comment by vladzcloudius** [comment]
@piodul I think it's more than acceptable.
The idea behind your series that put the receiver side of hints in the streaming class which has a static shares number of 300 (while query has 1000) was to ensure that the impact is limited by 33%.

Your results are in the expected range.

I'm ok with closing the issue.
@slivne

**Comment by slivne** [comment]
While the fix improved the performance - rethinking what we are doing here I still think we may have an issue here - we are replacing a node - so we are streaming information from other nodes (and will also replay the hints - which seems like we are doing double writes).

This is especially true when repair based operations are done - which effectivly repair between all the live replicas - so why should we send hints.

I think that when repair based operations are done  we should consider to clear the hints. 

Thoughts @glommer / @asias / @avikivity  ?

**Comment by avikivity** [comment]
I guess that's true. If we're starting to repair a node, we can clear any pending hints (in either directions).

On the other hand, sending hints is cheaper than repair. Maybe repair should start by sending all the pending hints.

**Comment by glommer** [comment]
The problem is that as far as I know hints are stored unconditionally on past hints: it is possible to have 100 hints for 100 overwrites to the same key, or hints for data that is not valid anymore.

Another issue is that if you want to replay all hints, depending on load conditions you can generate more hints in the process: as the load increases with the hint replay, you may generate more. There's no guarantee of convergence.

So my view is that as repair starts we should indeed just drop hints.

**Comment by vladzcloudius** [comment]
> The problem is that as far as I know hints are stored unconditionally on past hints: it is possible to have 100 hints for 100 overwrites to the same key, or hints for data that is not valid anymore.
> 
> Another issue is that if you want to replay all hints, depending on load conditions you can generate more hints in the process: as the load increases with the hint replay, you may generate more. There's no guarantee of convergence.
> 
> So my view is that as repair starts we should indeed just drop hints.

I also agree with @slivne on this.

**Comment by slivne** [comment]
@avikivity / @glommer we need to decide on the direction 
- dropping hints
- replaying them and differing the repair based operation till they have been replayed (at least some) 

I want to assign the work on this

**Comment by glommer** [comment]
As already stated, my view is drop hints.

**Comment by slivne** [comment]
So, lets use Avi's suggestion (https://github.com/scylladb/scylla/issues/4712#issuecomment-626602867) - lets start by replaying the hints and afterwords continue with repair.

**Comment by piodul** [comment]
@avikivity @slivne

I'd like to finally get around to implement the fix, but I'd also like to revisit the decision. I agree with Glauber and @vladzcloudius that dropping hints either before or after repair is a better idea then replaying them.

IMO, Glauber's observations about the nature of hints disqualify a solution where we wait until hints to a given node are sent. We have no guarantee that we get to a point when we have no hints - we might get stuck. Even if we only waited for hints which are present on disk _at the moment_ of repair start, we could end up with more hints being generated while we wait - so we are back to square one.

Moreover, hints are stored in a per-destination queue, but repair is done with at least keyspace granularity. If we want to replay hints for a particular keyspace or table only, we need to wait until hints to all other keyspaces/tables are sent too, which seems unnecessary.

Instead, I suggest that we mark a point at the current end of the hints queue and, during hints replay, ignore hints for the keyspace/table until we reach that point.

PS: Do we care about writes with CL=ANY? If we decide to drop hints, we may lose some writes done with this consistency.

**Comment by avikivity** [comment]
> @avikivity @slivne
> 
> I'd like to finally get around to implement the fix, but I'd also like to revisit the decision. I agree with Glauber and @vladzcloudius that dropping hints either before or after repair is a better idea then replaying them.
> 
> IMO, Glauber's observations about the nature of hints disqualify a solution where we wait until hints to a given node are sent. We have no guarantee that we get to a point when we have no hints - we might get stuck. Even if we only waited for hints which are present on disk _at the moment_ of repair start, we could end up with more hints being generated while we wait - so we are back to square one.

If a node is online, we should not add more hints for it. Our system does not hint slow nodes, only dead ones. So if we're attempting to repair, we should only need to replay the backlog, not catch up with new backlog.

> 
> Moreover, hints are stored in a per-destination queue, but repair is done with at least keyspace granularity. If we want to replay hints for a particular keyspace or table only, we need to wait until hints to all other keyspaces/tables are sent too, which seems unnecessary.
> 
> Instead, I suggest that we mark a point at the current end of the hints queue and, during hints replay, ignore hints for the keyspace/table until we reach that point.

I agree with that as a way of ensuring forward progress, but note that the hints queue for the nodes we repair should not grow.

> 
> PS: Do we care about writes with CL=ANY? If we decide to drop hints, we may lose some writes done with this consistency.

We should keep those hints. CL=ANY is pretty bad, but if we drop hints at any opportunity, we make it worthless.

**Comment by piodul** [comment]
@avikivity 

> If a node is online, we should not add more hints for it. Our system does not hint slow nodes, only dead ones. So if we're attempting to repair, we should only need to replay the backlog, not catch up with new backlog.

AFAIK, hints are generated in two cases: either when a replica is down, or when replica write times out (but CL is otherwise achieved). Therefore we do hint slow nodes which do not keep up with responding.

> I agree with that as a way of ensuring forward progress, but note that the hints queue for the nodes we repair should not grow.

Do you mean that the cluster needs to have enough spare capacity, so that writes do not timeout, in order to run repair? Under such assumption replaying hints before repair seems OK. Marking the point in the queue until which we will wait will protect us in case this assumption doesn't hold.

> We should keep those hints. CL=ANY is pretty bad, but if we drop hints at any opportunity, we make it worthless.

OK. In that case, replaying hints before repair seems to be a better solution. But...

...one more problem came to my mind: what about "orphaned" hints? Currently, if we have a hint in the queue towards X, but X is no longer its replica, then we send with CL=ALL to the new set of replicas. If some of them are down, then we will be blocked, because the current logic will keep retrying it.

**Comment by avikivity** [comment]
> @avikivity
> 
> > If a node is online, we should not add more hints for it. Our system does not hint slow nodes, only dead ones. So if we're attempting to repair, we should only need to replay the backlog, not catch up with new backlog.
> 
> AFAIK, hints are generated in two cases: either when a replica is down, or when replica write times out (but CL is otherwise achieved). Therefore we do hint slow nodes which do not keep up with responding.

We also have an additional mechanism where we wait for slow nodes. So these mechanisms are in conflict.

The theory behind it is that in a throughput oriented workload there will always be a slow node, so we'll always be generating hints and replaying them, which is wasted work. We should wait for the write to complete in the first place, if the node is up.

> 
> > I agree with that as a way of ensuring forward progress, but note that the hints queue for the nodes we repair should not grow.
> 
> Do you mean that the cluster needs to have enough spare capacity, so that writes do not timeout, in order to run repair? 

We can't make such an assumption (well, we can make it, but reality need not match the assumption).

> Under such assumption replaying hints before repair seems OK. Marking the point in the queue until which we will wait will protect us in case this assumption doesn't hold.

Yes. I still would like to understand how to reconcile the two mechanisms we have to handle slow nodes. The first one works from the point where we achieve CL until the node responds, the second starts working at timeout.

> 
> > We should keep those hints. CL=ANY is pretty bad, but if we drop hints at any opportunity, we make it worthless.
> 
> OK. In that case, replaying hints before repair seems to be a better solution. But...
> 
> ...one more problem came to my mind: what about "orphaned" hints? Currently, if we have a hint in the queue towards X, but X is no longer its replica, then we send with CL=ALL to the new set of replicas. If some of them are down, then we will be blocked, because the current logic will keep retrying it.

Good question. Perhaps we should replay to the set of nodes that are up. It will cause over-replay since we'll replay again when the down node is back up.

**Comment by vladzcloudius** [comment]
> We also have an additional mechanism where we wait for slow nodes. So these mechanisms are in conflict.

Not sure, which mechanisms you refer, @avikivity. As far as I know we don't have any such mechanism. If write to a replica failed (timed out) - that's it. Nothing except for a HH would happen till a repair in the context of this update.

So, it would help if you could explain which mechanism is in conflict with HH.

If you refer a write CL as a "mechanism" then I don't see where it's in conflict with HH since hints are generated only for replicas where write has failed and only if a CL has been achieved before that.
Namely if CL has been achieved and the rest of writes to replicas beyond the required for CL succeed - no hints will be generated.

As to HH being intended only for "node is DOWN" case - I'm also not sure what is this statement is based on. HH is actually pretty bad this this case because in this case it's likely going to fail storing most of hints because a local I/O backend is very likely not going to keep up.

And on the other hand it works decently well for "node is UP but some (few) writes to it fail from time to time".

> 
> The theory behind it is that in a throughput oriented workload there will always be a slow node, so we'll always be generating hints and replaying them, which is wasted work. We should wait for the write to complete in the first place, if the node is up.

We always wait for a write to complete before generating a hint (as I explained above). Hint would only be generated after we waited for a write to complete but it failed (plus the CL requirement has to be fulfilled).

> 
> > > I agree with that as a way of ensuring forward progress, but note that the hints queue for the nodes we repair should not grow.
> > 
> > 
> > Do you mean that the cluster needs to have enough spare capacity, so that writes do not timeout, in order to run repair?
> 
> We can't make such an assumption (well, we can make it, but reality need not match the assumption).
> 
> > Under such assumption replaying hints before repair seems OK. Marking the point in the queue until which we will wait will protect us in case this assumption doesn't hold.
> 
> Yes. I still would like to understand how to reconcile the two mechanisms we have to handle slow nodes. The first one works from the point where we achieve CL until the node responds, the second starts working at timeout.
> 
> > > We should keep those hints. CL=ANY is pretty bad, but if we drop hints at any opportunity, we make it worthless.
> > 
> > 
> > OK. In that case, replaying hints before repair seems to be a better solution. But...
> > ...one more problem came to my mind: what about "orphaned" hints? Currently, if we have a hint in the queue towards X, but X is no longer its replica, then we send with CL=ALL to the new set of replicas. If some of them are down, then we will be blocked, because the current logic will keep retrying it.
> 
> Good question. Perhaps we should replay to the set of nodes that are up. It will cause over-replay since we'll replay again when the down node is back up.

There is already a "drain" mode where we'd ignore failures and discard the segment - you can simply use it, @piodul

**Comment by gleb-cloudius** [comment]
> > @avikivity
> > > If a node is online, we should not add more hints for it. Our system does not hint slow nodes, only dead ones. So if we're attempting to repair, we should only need to replay the backlog, not catch up with new backlog.
> > 
> > 
> > AFAIK, hints are generated in two cases: either when a replica is down, or when replica write times out (but CL is otherwise achieved). Therefore we do hint slow nodes which do not keep up with responding.
> 
> We also have an additional mechanism where we wait for slow nodes. So these mechanisms are in conflict.
> 

What is this mechanism? Do you mean write throttling? If yes, then I do not see any conflict between it and HH. Throttling mechanism postpones an already ready reply until a request is complete from coordinator's POV. The completion may happen in two ways: all replicas relied or timeout is reached and a hint is written for those that didn't.

**Comment by piodul** [comment]
I created a separate issue about the idea of replaying hints before repair, let's move the discussion there: https://github.com/scylladb/scylla/issues/8102

**Comment by piodul** [comment]
- We agreed that the throughput drop during hints replay + repair during node replacement is much better and acceptable now
- The discussion about replaying hints before repair was moved to another issue

I see nothing left to address here, so I'm closing this issue.
